plugins/chat/lib/chat/parsed_mentions.rb

```diff
@@ -51,8 +51,9 @@ def direct_mentions
     end
 
     def group_mentions
-      mentionable_groups_ids = mentionable_groups.pluck(:id)
-      chat_users.includes(:groups).joins(:groups).where("groups.id IN (?)", mentionable_groups_ids)
+      group_ids = groups_to_mention.pluck(:id)
+      group_user_ids = GroupUser.where(group_id: group_ids).pluck(:user_id)
```

Comment:
If we do not need the intermediate IDs queried into Ruby arrays, we can use `#select` to avoid extra queries and allocations:

```suggestion
      group_ids = groups_to_mention.select(:id)
      group_user_ids = GroupUser.where(group_id: group_ids).select(:user_id)
      chat_users.where(id: group_user_ids)
```

Reply:
Hm, this way Active Record will be making a heavy join of the `users` and `groups` tables, exactly what it did before a change I made in this PR. Avoiding that join by forcing AR to do two lightweight queries instead is the central point of this optimization:

> This PR eliminates one more heavy join by forcing Active Record to do two queries instead.

I'm going to merge it to look how good the impact on real data is going to be (I'm definitely seeing improvement on my local instance). We can discuss and reconsider it after in any case.

Reply:
I merged it and confirmed that this seriously improved performance on real data, but still not enough. Here is a follow-up with further improvements – https://github.com/discourse/discourse/pull/24185.